src/main/java/com/mongodb/hibernate/translate/attachment/Attachment.java

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+import static java.lang.String.format;
+
+import com.mongodb.hibernate.internal.VisibleForTesting;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * Manages global state while translating SQL AST via {@link org.hibernate.sql.ast.SqlAstWalker}.
+ *
+ * <p>{@code SqlAstWalker} is based on {@code visitor} pattern without returning value from the various visitor methods.
+ * The original goal of {@code SqlAstWalker} is for concatenating SQL string internally; but when translating SQL AST
+ * into MQL, it is highly desirable to return value (e.g. {@link com.mongodb.hibernate.translate.mongoast.AstNode}). To
+ * avoid such dilemma, some attachment/detachment mechanism is implemented as follows:
+ *
+ * <ul>
+ *   <li>when some visitor method needs to return some value, it attaches the value by invoking
+ *       {@link #attach(AttachmentKey, Object)}.
+ *   <li>when some coordinating detacher needs to access the value attached above, it invokes
+ *       {@link #expect(AttachmentKey, Runnable)}; internally the {@code Runnable} will attach value by the above
+ *       attacher.
+ *   <li>both attacher and detacher should provide identical {@link AttachmentKey} to ensure data transferring safety.
+ * </ul>
+ *
+ * @see org.hibernate.sql.ast.SqlAstWalker
+ * @see AttachmentKey
+ */
+@SuppressWarnings("unchecked")
+public final class Attachment {
+
+    private @Nullable AttachmentKey<?> attachmentKey;
+    private @Nullable Object attachedValue;
+
+    @Override
+    public String toString() {
+        return "Attachment{" + "attachmentKey=" + attachmentKey + ", attachedValue=" + attachedValue + "}";
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.AccessModifier.PRIVATE)
+    boolean isBlank() {
+        return this.attachmentKey == null && this.attachedValue == null;
+    }
+
+    /**
+     * Detaches the attached value by invoking some visitor method which returns {@code void} but internally the method
+     * is expected to attach some value by invoking the other {@link #attach(AttachmentKey, Object)} method.
+     *
+     * @param attachmentKey type safe attachment type enum; detacher and attacher should share the identical
+     *     {@code AttachmentKey}; never null
+     * @param attacher some {@link Runnable} wrapper of visitor method which is supposed to attach some value
+     *     internally; never {@code null}
+     * @return the value attached by the attacher; after this method returns, the previous state will be restored; never
+     *     {@code null}
+     * @param <T> the expected Java generic type
+     */
+    public <T> T expect(AttachmentKey<T> attachmentKey, Runnable attacher) {
+        var previousAttachmentKey = this.attachmentKey;
```

Comment:
Isn't it true that `previousAttachmentKey` is always `null`, provided that `Attachment.expect` is not called concurrently with itself?

I suspect, that all we need is to assert here that `this.attachmentKey` is `null`, and we don't need `previousAttachmentKey`.

Reply:
no, previouslyAttachmentKey could be anything. We use it to cover such common AST visiting pattern: one visitor method calls another one, which calls another one, ..... AST tree depth could be ad-hoc, sub query is a good example.

That is why I originally used `Stack` to record the tree traversal (post-order for we attach value bottom-up), but later on I realize Jeff's implementation has covered the case of backtracking by reverting back previous states.

So suppose in a deeply nested visitor method calling, all the "previous" state will be stored in their corresponding statck trace local variables. This seems an elegant solution (though it takes some getting used to and deep understanding, which could confuse new coder invariably).